From 2c9b2c9cac5ec2b659f2ccd77de6188c7d3b7d19 Mon Sep 17 00:00:00 2001
From: Amelie Delaunay <amelie.delaunay@st.com>
Date: Tue, 14 Jan 2020 11:15:27 +0100
Subject: [PATCH] usb: dwc2: keep the usb stack informed of SetPortFeature
 failure while Host

In case port connect status is disabled due to a device disconnection but
the device is still physically plugged so the core is still is Host mode,
USB stack needs to know that the SetPortFeature fails. It is also the case
when the core is configured as Host only.

So, when the device is disconnected but the core is still in Host mode,
let's the driver manage the SetPortFeature asked by the USB stack. Then the
port is not completely stucked and the device can be re-enumerated on plug.

Change-Id: I2616750b8b276e82b61ea67b459a7784e02adf86
Signed-off-by: Amelie Delaunay <amelie.delaunay@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/157591
Reviewed-by: CITOOLS <smet-aci-reviews@lists.codex.cro.st.com>
Reviewed-by: CIBUILD <smet-aci-builds@lists.codex.cro.st.com>
Reviewed-by: Fabrice GASNIER <fabrice.gasnier@st.com>
---
 drivers/usb/dwc2/hcd.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/dwc2/hcd.c b/drivers/usb/dwc2/hcd.c
index 5622083c1aca..40be338278bc 100644
--- a/drivers/usb/dwc2/hcd.c
+++ b/drivers/usb/dwc2/hcd.c
@@ -3621,7 +3621,8 @@ static int dwc2_hcd_hub_control(struct dwc2_hsotg *hsotg, u16 typereq,
 		if (wvalue != USB_PORT_FEAT_TEST && (!windex || windex > 1))
 			goto error;
 
-		if (!hsotg->flags.b.port_connect_status) {
+		if (!hsotg->flags.b.port_connect_status &&
+		    !dwc2_is_host_mode(hsotg)) {
 			/*
 			 * The port is disconnected, which means the core is
 			 * either in device mode or it soon will be. Just
-- 
2.39.5

